idx/memory/memory.go

```diff
@@ -148,79 +163,179 @@ func (ids IdSet) String() string {
+func (t *TagIndex) addTagId(name, value uintptr, id schema.MKey) {
 	ti := *t
 	if _, ok := ti[name]; !ok {
 		ti[name] = make(TagValues)
+		interning.IdxIntern.IncRefCnt(name)
 	}
 	if _, ok := ti[name][value]; !ok {
 		ti[name][value] = make(IdSet)
+		interning.IdxIntern.IncRefCnt(value)
 	}
 	ti[name][value][id] = struct{}{}
 }
 
-func (t *TagIndex) delTagId(name, value string, id schema.MKey) {
+// addTagIdForName does the same as addTagId, but for the special
+// that "name" which indexes the metric's name value.
+// it needs to be treated differently because, unlike normal tags,
+// it is not guaranteed that the name value has already been
+// interned at this point
+func (t *TagIndex) addTagIdForName(name string, id schema.MKey) {
+	ti := *t
+
+	var namePtr, valuePtr uintptr
+	namePtr, _ = interning.IdxIntern.GetPtrFromByte([]byte("name"))
+	newlyInterned := false
+	if namePtr == 0 {
+		namePtr, _ = interning.IdxIntern.AddOrGet([]byte("name"), false)
+		newlyInterned = true
+	}
+
+	if _, ok := ti[namePtr]; !ok {
+		ti[namePtr] = make(TagValues)
+		if !newlyInterned {
+			interning.IdxIntern.IncRefCnt(namePtr)
+		}
+	}
+
+	valuePtr, _ = interning.IdxIntern.GetPtrFromByte([]byte(name))
+	newlyInterned = false
+	if valuePtr == 0 {
+		valuePtr, _ = interning.IdxIntern.AddOrGet([]byte(name), false)
+		newlyInterned = true
+	}
+
+	if _, ok := ti[namePtr][valuePtr]; !ok {
+		ti[namePtr][valuePtr] = make(IdSet)
+		if !newlyInterned {
+			interning.IdxIntern.IncRefCnt(valuePtr)
+		}
+	}
+
+	ti[namePtr][valuePtr][id] = struct{}{}
+}
+
+func (t *TagIndex) delTagId(name, value uintptr, id schema.MKey) {
 	ti := *t
 
 	delete(ti[name][value], id)
 
 	if len(ti[name][value]) == 0 {
 		delete(ti[name], value)
+		interning.IdxIntern.Delete(value)
 		if len(ti[name]) == 0 {
 			delete(ti, name)
+			interning.IdxIntern.Delete(name)
+		}
+	}
+}
+
+func (t *TagIndex) delTagIdForName(name string, id schema.MKey) {
+	ti := *t
+
+	namePtr, err := interning.IdxIntern.GetPtrFromByte([]byte("name"))
+	if err != nil || namePtr == 0 {
+		log.Error("memory-idx: Failed to retrieve interned string for 'name' key: ", err)
+		internError.Inc()
+	}
+
+	valuePtr, err := interning.IdxIntern.GetPtrFromByte([]byte(name))
+	if err != nil || valuePtr == 0 {
+		log.Error("memory-idx: Failed to retrieve interned string for 'name' value: ", err)
+		internError.Inc()
+	}
+
+	delete(ti[namePtr][valuePtr], id)
+
+	if len(ti[namePtr][valuePtr]) == 0 {
+		delete(ti[namePtr], valuePtr)
+		interning.IdxIntern.Delete(valuePtr)
+		if len(ti[namePtr]) == 0 {
+			delete(ti, namePtr)
+			interning.IdxIntern.Delete(namePtr)
 		}
 	}
 }
 
-// org id -> nameWithTags -> Set of references to schema.MetricDefinition
-// nameWithTags is the name plus all tags in the <name>;<tag>=<value>... format.
-type defByTagSet map[uint32]map[string]map[*schema.MetricDefinition]struct{}
+func (t TagIndex) idHasTag(id schema.MKey, tag, value uintptr) bool {
+	_, ok := t[tag][value][id]
+	return ok
```

Comment:
this can panic if one of the intermediate maps doesn't exist.
maybe we should look up the metricdef by id ?

Reply:
i don't think this should panic because when a non-existent key is looked up the return value is `nil`. `nil` is a valid map for reading (although not for writing).
https://play.golang.org/p/ZaYEovPqMuF

Reply:
huh. whoa!